lib/_emerge/resolver/circular_dependency.py

```diff
@@ -132,9 +132,15 @@ def _find_suggestions(self):
             for ppkg, atom in parent_atoms:
                 if ppkg == parent:
                     changed_parent = ppkg
-                    parent_atom = atom.unevaluated_atom
+                    parent_atom = atom
                     break
 
+            if parent_atom.package:
+                parent_atom = parent_atom.unevaluated_atom
+            else:
+                # SonameAtom is always unconditional.
+                continue
```

Comment:
The bug was actually triggered by this cycle which only occurs with emerge --ignore-soname-deps=n because gmp does not specify a dependency on gcc:
```
(dev-libs/gmp-6.3.0-2:0/10.4::gentoo, binary scheduled for merge to '/tmp/testroot/') depends on
 (sys-devel/gcc-13.2.1_p20230826-2:13/13::gentoo, binary scheduled for merge to '/tmp/testroot/') (runtime)
  (dev-libs/gmp-6.3.0-2:0/10.4::gentoo, binary scheduled for merge to '/tmp/testroot/') (runtime_slot_op)
```
Also, in theory you can break the cycle by disabling USE=cxx for gmp, but since the dependency is unspecified it's kind of an odd thing to do.

Reply:
Adjusted comments and commit message for this.